File: drills/elixir/solutions/nc66_house_robber__memoised.exs
defmodule Solution do
  def rob(nums) do
    {best, _memo} = from(0, List.to_tuple(nums), length(nums), %{})
    best
  end

  # The same choice written as a recursion from the front: rob this house and
  # skip the next, or skip this one. Exponential without the cache and linear
  # with it -- which is the lesson, since the rolling pair hides that the
  # problem ever had a tree of choices at all.
  defp from(index, _houses, count, memo) when index >= count, do: {0, memo}

  defp from(index, houses, count, memo) do
    case Map.fetch(memo, index) do
      {:ok, cached} ->
        {cached, memo}

      :error ->
        {taken, memo} = from(index + 2, houses, count, memo)
        {skipped, memo} = from(index + 1, houses, count, memo)
        best = max(elem(houses, index) + taken, skipped)
        {best, Map.put(memo, index, best)}
    end
  end
end

File: drills/go/solutions/nc66_house_robber__memoised.go
package main

func rob(nums []int) int {
	memo := map[int]int{}
	var from func(i int) int
	from = func(i int) int {
		if i >= len(nums) {
			return 0
		}
		if v, ok := memo[i]; ok {
			return v
		}
		memo[i] = max(nums[i]+from(i+2), from(i+1))
		return memo[i]
	}
	return from(0)
}
